Re-enable RUST_BACKTRACE

Is the old implementation faster again? Does it need to be opt-in from the end user?

Example ICE that I just stumbled upon([but was already found by others](https://github.com/rust-lang/rust/issues/43197)), re durka's [comment](https://github.com/integer32llc/rust-playground/issues/132#issuecomment-317895791):

```rust
pub const X: i64 = std::i64::MAX + 1;
fn main() {
    println!("{}", X);
}
```
([^ playground link](https://play.rust-lang.org/?gist=bcef2b956253bbd4ed6afd21ea6714a0&version=nightly))

<details>
 <summary>
output
</summary>
<div style="white-space: pre-wrap;">
<pre style="white-space: pre-wrap;">
<div style="white-space: pre-wrap;">
   Compiling playground v0.0.1 (file:///playground)
warning: constant evaluation error: attempt to add with overflow. This will become a HARD ERROR in the future
 --> src/main.rs:1:20
  |
1 | pub const X: i64 = std::i64::MAX + 1;
  |                    ^^^^^^^^^^^^^^^^^
  |
  = note: #[warn(const_err)] on by default

error[E0080]: constant evaluation error
 --> src/main.rs:1:20
  |
1 | pub const X: i64 = std::i64::MAX + 1;
  |                    ^^^^^^^^^^^^^^^^^ attempt to add with overflow

error: internal compiler error: /checkout/src/librustc_trans/mir/constant.rs:396: _1 not initialized
 --> src/main.rs:3:20
  |
3 |     println!("{}", X);
  |                    ^

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.21.0-nightly (52a330969 2017-07-27) running on x86_64-unknown-linux-gnu

thread 'rustc' panicked at 'Box<Any>', /checkout/src/librustc_errors/lib.rs:438:8
note: Run with `RUST_BACKTRACE=1` for a backtrace.

error: Could not compile `playground`.

To learn more, run the command again with --verbose.
</div>
</pre>
If someone knows how to make github wrap long lines in the above output, please let me know. 
</div>
</details>

------

[playground link](https://gist.github.com/eaae31f2fca316b87823bff06441fee8) for this ICE with [durka's workaround](https://github.com/integer32llc/rust-playground/issues/132#issuecomment-317895791) for displaying ICE stacktraces - thank you btw!
<details>
<summary>
code for the above
</summary>
<pre>
use std::io::Write;
use std::fs::File;
use std::process::Command;

fn main() {
    let code = stringify!(
        pub const X: i64 = std::i64::MAX + 1;
fn main() {
    println!("{}", X);
}

    );

    writeln!(File::create("foo.rs").unwrap(), "{}", code).unwrap();

    let cmd = Command::new("rustc")
        .env("RUST_BACKTRACE", "full")
        .arg("+nightly")
        .arg("foo.rs")
        .status()
        .unwrap();
    println!("exit code {}", cmd);
}

</pre>
</details>
<details>

<summary>
output for the above
</summary>
<pre>
   Compiling playground v0.0.1 (file:///playground)
    Finished dev [unoptimized + debuginfo] target(s) in 0.86 secs
     Running `target/debug/playground`
warning: constant evaluation error: attempt to add with overflow. This will become a HARD ERROR in the future
 --> foo.rs:1:21
  |
1 | pub const X : i64 = std :: i64 :: MAX + 1 ; fn main (  ) {
  |                     ^^^^^^^^^^^^^^^^^^^^^
  |
  = note: #[warn(const_err)] on by default

error[E0080]: constant evaluation error
 --> foo.rs:1:21
  |
1 | pub const X : i64 = std :: i64 :: MAX + 1 ; fn main (  ) {
  |                     ^^^^^^^^^^^^^^^^^^^^^ attempt to add with overflow

error: internal compiler error: /checkout/src/librustc_trans/mir/constant.rs:396: _1 not initialized
 --> foo.rs:2:20
  |
2 | println ! ( "{}" , X ) ; }
  |                    ^

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.21.0-nightly (52a330969 2017-07-27) running on x86_64-unknown-linux-gnu

note: run with `RUST_BACKTRACE=1` for a backtrace

thread 'rustc' panicked at 'Box<Any>', /checkout/src/librustc_errors/lib.rs:438:8
stack backtrace:
   0:     0x7f0cbc947db3 - std::sys::imp::backtrace::tracing::imp::unwind_backtrace::h1e2c58aa3ff6d86d
                               at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1:     0x7f0cbc942494 - std::sys_common::backtrace::_print::h021c974b848276a0
                               at /checkout/src/libstd/sys_common/backtrace.rs:71
   2:     0x7f0cbc955c33 - std::panicking::default_hook::{{closure}}::heee72575d0ec5077
                               at /checkout/src/libstd/sys_common/backtrace.rs:60
                               at /checkout/src/libstd/panicking.rs:380
   3:     0x7f0cbc955928 - std::panicking::default_hook::hfd71e7dd65b0a86c
                               at /checkout/src/libstd/panicking.rs:390
   4:     0x7f0cbc956137 - std::panicking::rust_panic_with_hook::h4eba8d7e50a7fffa
                               at /checkout/src/libstd/panicking.rs:610
   5:     0x7f0cbb84f878 - std::panicking::begin_panic::h48e4cf472e4fad07
   6:     0x7f0cbb87b668 - rustc::session::opt_span_bug_fmt::{{closure}}::haf7ee303771aa9d5
   7:     0x7f0cbb87b4c7 - rustc::session::span_bug_fmt::haa62c77dec292770
   8:     0x7f0cbb91a0f3 - rustc_trans::mir::constant::MirConstContext::const_lvalue::h1dbacf146fcd1592
   9:     0x7f0cbb91aeec - rustc_trans::mir::constant::MirConstContext::const_rvalue::h1269edfb1e998250
  10:     0x7f0cbb91898c - rustc_trans::mir::constant::MirConstContext::trans::hfa9f803bd5275c21
  11:     0x7f0cbb91e7bf - rustc_trans::mir::constant::<impl rustc_trans::mir::MirContext<'a, 'tcx>>::trans_constant::h5da8b49ad7fff201
  12:     0x7f0cbb921fa1 - rustc_trans::mir::operand::<impl rustc_trans::mir::MirContext<'a, 'tcx>>::trans_operand::hec9f10a4034bfc20
  13:     0x7f0cbb923684 - rustc_trans::mir::rvalue::<impl rustc_trans::mir::MirContext<'a, 'tcx>>::trans_rvalue_operand::h865cdcdd7c0b92cf
  14:     0x7f0cbb90c713 - rustc_trans::mir::trans_mir::h559e0cace6be4e5a
  15:     0x7f0cbb92ff33 - rustc_trans::trans_item::TransItem::define::h8e122b431d2e3ba3
  16:     0x7f0cbb8d4bed - rustc_trans::base::trans_crate::module_translation::hc4a2a7b27dcd9f96
  17:     0x7f0cbb8cecfe - rustc_trans::base::trans_crate::h15091f3336a483e1
  18:     0x7f0cbcd02902 - rustc_driver::driver::phase_4_translate_to_llvm::h911c0cc67ac5141c
  19:     0x7f0cbcccf79a - rustc_driver::driver::compile_input::{{closure}}::h81c7e0ebddde0b7a
  20:     0x7f0cbcd0174e - rustc_driver::driver::phase_3_run_analysis_passes::{{closure}}::hea490b5adce91e62
  21:     0x7f0cbcce3823 - rustc_driver::driver::phase_3_run_analysis_passes::h1210be0da5a54c32
  22:     0x7f0cbcccdb30 - rustc_driver::driver::compile_input::hbf08f2b62431a302
  23:     0x7f0cbcd12f65 - rustc_driver::run_compiler::h7b5ef3de0e76a53e
  24:     0x7f0cbcc2c477 - std::sys_common::backtrace::__rust_begin_short_backtrace::he953d71f28957e91
  25:     0x7f0cbc98567c - __rust_maybe_catch_panic
                               at /checkout/src/libpanic_unwind/lib.rs:98
  26:     0x7f0cbcc63e60 - <F as alloc::boxed::FnBox<A>>::call_box::hb0c50b1ae5590427
  27:     0x7f0cbc954b2b - std::sys::imp::thread::Thread::new::thread_start::h8cc7df731d74d00c
                               at /checkout/src/liballoc/boxed.rs:692
                               at /checkout/src/libstd/sys_common/thread.rs:21
                               at /checkout/src/libstd/sys/unix/thread.rs:84
  28:     0x7f0cb7aec6b9 - start_thread
  29:     0x7f0cbc6023dc - clone
  30:                0x0 - <unknown>

Standard Output
exit code exit code: 101
</pre>
</details>

--- a/tests/spec/features/backtrace_spec.rb
+++ b/tests/spec/features/backtrace_spec.rb
@@ -1,26 +1,45 @@
 require 'spec_helper'
 require 'support/editor'
+require 'support/playground_actions'
 
 RSpec.feature "A backtrace is shown for certain errors", type: :feature, js: true do
-  before { pending "Backtracing has a large performance penalty" }
+  include PlaygroundActions
 
   before do
     visit '/'
     editor.set(code)
-    within('.header') { click_on("Run") }
   end
 
-  scenario "a stack trace is shown" do
-    within('.output-stderr') do
-      expect(page).to have_content 'stack backtrace:'
-      expect(page).to have_content 'rust_begin_unwind'
+  context "backtraces are enabled" do
+    before do
+      in_advanced_options_menu { choose 'enabled' }
+      within('.header') { click_on("Run") }
+    end
+
+    scenario "a backtrace is shown" do
+      within('.output-stderr') do
+        expect(page).to have_content 'stack backtrace:'
+        expect(page).to have_content 'rust_begin_unwind'
+      end
+    end
+
+    scenario "filenames link to that line of code" do
+      within('.output-stderr') do
+        expect(page).to have_link('main.rs:2')
+        expect(page).to have_link('main.rs:6')
+      end
     end
   end
 
-  scenario "filenames link to that line of code" do
-    within('.output-stderr') do
-      expect(page).to have_link('main.rs:2')
-      expect(page).to have_link('main.rs:6')
+  context "backtraces are disabled" do
+    before do
+      within('.header') { click_on("Run") }
+    end
+
+    scenario "the backtrace suggestion is a link" do
+      within('.output-stderr') do
+        expect(page).to have_link(text: /Run with .* a backtrace/)
+      end
     end
   end
 

--- a/ui/frontend/AdvancedOptionsMenu.tsx
+++ b/ui/frontend/AdvancedOptionsMenu.tsx
@@ -1,19 +1,22 @@
 import React, { Fragment } from 'react';
 import { connect } from 'react-redux';
 
-import { changeEdition } from './actions';
+import { changeBacktrace, changeEdition } from './actions';
 import { changeNightlyEdition } from './actions';
 import { Either as EitherConfig } from './ConfigElement';
 import MenuGroup from './MenuGroup';
 import { State } from './reducers';
-import { getEditionSet, isEditionAvailable } from './selectors';
-import { Edition } from './types';
+import { getBacktraceSet, getEditionSet, isEditionAvailable } from './selectors';
+import { Backtrace, Edition } from './types';
 
 interface AdvancedOptionsMenuProps {
   edition: Edition;
   isEditionSet: boolean;
   isEditionAvailable: boolean;
   changeEdition: (_: Edition) => any;
+  backtrace: Backtrace;
+  isBacktraceSet: boolean;
+  changeBacktrace: (_: Backtrace) => any;
 }
 
 const AdvancedOptionsMenu: React.SFC<AdvancedOptionsMenuProps> = props => (
@@ -28,6 +31,15 @@ const AdvancedOptionsMenu: React.SFC<AdvancedOptionsMenuProps> = props => (
         isNotDefault={props.isEditionSet}
         onChange={props.changeEdition} />
       {!props.isEditionAvailable && <EditionAside />}
+
+      <EitherConfig
+        id="backtrace"
+        name="Backtrace"
+        a={Backtrace.Disabled}
+        b={Backtrace.Enabled}
+        value={props.backtrace}
+        isNotDefault={props.isBacktraceSet}
+        onChange={props.changeBacktrace} />
     </MenuGroup>
   </Fragment>
 );
@@ -43,10 +55,13 @@ const mapStateToProps = (state: State) => ({
   isEditionSet: getEditionSet(state),
   edition: state.configuration.edition,
   isEditionAvailable: isEditionAvailable(state),
+  isBacktraceSet: getBacktraceSet(state),
+  backtrace: state.configuration.backtrace,
 });
 
 const mapDispatchToProps = ({
   changeEdition: changeNightlyEdition,
+  changeBacktrace,
 });
 
 export default connect(mapStateToProps, mapDispatchToProps)(AdvancedOptionsMenu);

--- a/ui/frontend/actions.ts
+++ b/ui/frontend/actions.ts
@@ -6,6 +6,7 @@ import { getCrateType, isEditionAvailable, runAsTest } from './selectors';
 import State from './state';
 import {
   AssemblyFlavor,
+  Backtrace,
   Channel,
   DemangleAssembly,
   Edition,
@@ -51,6 +52,7 @@ export enum ActionType {
   ChangeProcessAssembly = 'CHANGE_PROCESS_ASSEMBLY',
   ChangeMode = 'CHANGE_MODE',
   ChangeEdition = 'CHANGE_EDITION',
+  ChangeBacktrace = 'CHANGE_BACKTRACE',
   ChangeFocus = 'CHANGE_FOCUS',
   ExecuteRequest = 'EXECUTE_REQUEST',
   ExecuteSucceeded = 'EXECUTE_SUCCEEDED',
@@ -130,6 +132,9 @@ export const changeNightlyEdition = (edition: Edition): ThunkAction => dispatch 
   dispatch(changeChannel(Channel.Nightly));
   dispatch(changeEdition(edition));
 };
+
+export const changeBacktrace = (backtrace: Backtrace) =>
+  createAction(ActionType.ChangeBacktrace, { backtrace });
 
 export const changeFocus = focus =>
   createAction(ActionType.ChangeFocus, { focus });
@@ -569,6 +574,7 @@ export type Action =
   | ReturnType<typeof changeKeybinding>
   | ReturnType<typeof changeMode>
   | ReturnType<typeof changeEdition>
+  | ReturnType<typeof changeBacktrace>
   | ReturnType<typeof changeOrientation>
   | ReturnType<typeof changeTheme>
   | ReturnType<typeof requestExecute>

--- a/ui/frontend/actions.ts
+++ b/ui/frontend/actions.ts
@@ -136,6 +136,11 @@ export const changeNightlyEdition = (edition: Edition): ThunkAction => dispatch 
 export const changeBacktrace = (backtrace: Backtrace) =>
   createAction(ActionType.ChangeBacktrace, { backtrace });
 
+export const reExecuteWithBacktrace = (): ThunkAction => dispatch => {
+  dispatch(changeBacktrace(Backtrace.Enabled));
+  dispatch(performExecute());
+};
+
 export const changeFocus = focus =>
   createAction(ActionType.ChangeFocus, { focus });
 

--- a/ui/frontend/actions.ts
+++ b/ui/frontend/actions.ts
@@ -189,6 +189,7 @@ interface ExecuteRequestBody {
   tests: boolean;
   code: string;
   edition?: string;
+  backtrace: boolean;
 }
 
 export function performExecute(): ThunkAction {
@@ -200,8 +201,9 @@ export function performExecute(): ThunkAction {
     const { code, configuration: { channel, mode, edition } } = state;
     const crateType = getCrateType(state);
     const tests = runAsTest(state);
+    const backtrace = state.configuration.backtrace === Backtrace.Enabled;
 
-    const body: ExecuteRequestBody = { channel, mode, crateType, tests, code };
+    const body: ExecuteRequestBody = { channel, mode, crateType, tests, code, backtrace };
     if (isEditionAvailable(state)) {
       body.edition = edition;
     }
@@ -235,6 +237,7 @@ function performCompile(target, { request, success, failure }): ThunkAction {
     } } = state;
     const crateType = getCrateType(state);
     const tests = runAsTest(state);
+    const backtrace = state.configuration.backtrace === Backtrace.Enabled;
     const body: CompileRequestBody = {
       channel,
       mode,
@@ -245,6 +248,7 @@ function performCompile(target, { request, success, failure }): ThunkAction {
       assemblyFlavor,
       demangleAssembly,
       processAssembly,
+      backtrace,
     };
     if (isEditionAvailable(state)) {
       body.edition = edition;

--- a/ui/frontend/highlighting.ts
+++ b/ui/frontend/highlighting.ts
@@ -11,7 +11,12 @@ export function configureRustErrors({ gotoPosition, getChannel }) {
       },
     },
     'error-location': /-->.*\n/,
-    'stack-trace-location': /at \/playground.*\n/,
+    'backtrace': {
+      pattern: /at src\/.*\n/,
+      inside: {
+        'backtrace-location': /src\/main.rs:(\d+)/,
+      },
+    },
   };
 
   Prism.hooks.add('wrap', env => {
@@ -37,8 +42,8 @@ export function configureRustErrors({ gotoPosition, getChannel }) {
       env.attributes['data-line'] = line;
       env.attributes['data-col'] = col;
     }
-    if (env.type === 'stack-trace-location') {
-      const errorMatch = /main.rs:(\d+)/.exec(env.content);
+    if (env.type === 'backtrace-location') {
+      const errorMatch = /:(\d+)/.exec(env.content);
       const [_, line] = errorMatch;
       env.tag = 'a';
       env.attributes.href = '#';
@@ -48,7 +53,7 @@ export function configureRustErrors({ gotoPosition, getChannel }) {
   });
 
   Prism.hooks.add('after-highlight', env => {
-    const links = env.element.querySelectorAll('.error-location, .stack-trace-location');
+    const links = env.element.querySelectorAll('.error-location, .backtrace-location');
     Array.from(links).forEach((link: HTMLAnchorElement) => {
       const { line, col } = link.dataset;
       link.onclick = e => {

--- a/ui/frontend/highlighting.ts
+++ b/ui/frontend/highlighting.ts
@@ -1,6 +1,6 @@
 import Prism from 'prismjs';
 
-export function configureRustErrors({ gotoPosition, getChannel }) {
+export function configureRustErrors({ gotoPosition, reExecuteWithBacktrace, getChannel }) {
   Prism.languages.rust_errors = { // eslint-disable-line camelcase
     'warning': /^warning:.*$/m,
     'error': {
@@ -17,6 +17,7 @@ export function configureRustErrors({ gotoPosition, getChannel }) {
         'backtrace-location': /src\/main.rs:(\d+)/,
       },
     },
+    'backtrace-enable': /Run with `RUST_BACKTRACE=1` for a backtrace/,
   };
 
   Prism.hooks.add('wrap', env => {
@@ -42,6 +43,11 @@ export function configureRustErrors({ gotoPosition, getChannel }) {
       env.attributes['data-line'] = line;
       env.attributes['data-col'] = col;
     }
+    if (env.type === 'backtrace-enable') {
+      env.tag = 'a';
+      env.attributes.href = '#';
+      env.attributes['data-backtrace-enable'] = 'true';
+    }
     if (env.type === 'backtrace-location') {
       const errorMatch = /:(\d+)/.exec(env.content);
       const [_, line] = errorMatch;
@@ -59,6 +65,14 @@ export function configureRustErrors({ gotoPosition, getChannel }) {
       link.onclick = e => {
         e.preventDefault();
         gotoPosition(line, col);
+      };
+    });
+
+    const backtraceEnablers = env.element.querySelectorAll('.backtrace-enable');
+    Array.from(backtraceEnablers).forEach((link: HTMLAnchorElement) => {
+      link.onclick = e => {
+        e.preventDefault();
+        reExecuteWithBacktrace();
       };
     });
   });

--- a/ui/frontend/index.tsx
+++ b/ui/frontend/index.tsx
@@ -8,13 +8,14 @@ import persistState from 'redux-localstorage';
 import thunk, { ThunkDispatch } from 'redux-thunk';
 import * as url from 'url';
 
-import { Action, gotoPosition, performCratesLoad, performVersionsLoad } from './actions';
+import { Action, gotoPosition, performCratesLoad, performVersionsLoad, reExecuteWithBacktrace } from './actions';
 import { configureRustErrors } from './highlighting';
 import { deserialize, serialize } from './local_storage';
 import PageSwitcher from './PageSwitcher';
 import playgroundApp from './reducers';
 import { State } from './reducers';
 import Router from './Router';
+import { Backtrace } from './types';
 
 const baseUrl = url.resolve(window.location.href, '/');
 
@@ -31,6 +32,7 @@ const store = createStore(playgroundApp, initialState, enhancers);
 
 configureRustErrors({
   gotoPosition: (line, col) => store.dispatch(gotoPosition(line, col)),
+  reExecuteWithBacktrace: () => store.dispatch(reExecuteWithBacktrace()),
   getChannel: () => store.getState().configuration.channel,
 });
 

--- a/ui/frontend/reducers/configuration.ts
+++ b/ui/frontend/reducers/configuration.ts
@@ -1,6 +1,7 @@
 import { Action, ActionType } from '../actions';
 import {
   AssemblyFlavor,
+  Backtrace,
   Channel,
   DemangleAssembly,
   Edition,
@@ -22,6 +23,7 @@ export interface State {
   channel: Channel;
   mode: Mode;
   edition: Edition;
+  backtrace: Backtrace;
 }
 
 export const DEFAULT: State = {
@@ -36,6 +38,7 @@ export const DEFAULT: State = {
   channel: Channel.Stable,
   mode: Mode.Debug,
   edition: Edition.Rust2015,
+  backtrace: Backtrace.Disabled,
 };
 
 export default function configuration(state = DEFAULT, action: Action): State {
@@ -75,6 +78,8 @@ export default function configuration(state = DEFAULT, action: Action): State {
       }
       return { ...state, edition: action.edition };
     }
+    case ActionType.ChangeBacktrace:
+      return { ...state, backtrace: action.backtrace };
     default:
       return state;
   }

--- a/ui/frontend/selectors/index.ts
+++ b/ui/frontend/selectors/index.ts
@@ -3,7 +3,7 @@ import { createSelector } from 'reselect';
 import * as url from 'url';
 
 import { State } from '../reducers';
-import { Channel, Edition } from '../types';
+import { Backtrace, Channel, Edition } from '../types';
 
 const getCode = state => state.code;
 
@@ -53,12 +53,19 @@ export const getChannelLabel = (state: State) => {
   return `${channel}`;
 };
 
-export const getAdvancedOptionsSet = (state: State) => (
-  getEditionSet(state)
-);
-
 export const getEditionSet = (state: State) => (
   state.configuration.edition !== Edition.Rust2015
+);
+
+export const getBacktraceSet = (state: State) => (
+  state.configuration.backtrace !== Backtrace.Disabled
+);
+
+export const getAdvancedOptionsSet = createSelector(
+  getEditionSet, getBacktraceSet,
+  (editionSet, backtraceSet) => (
+    editionSet || backtraceSet
+  ),
 );
 
 const baseUrlSelector = (state: State) =>

--- a/ui/frontend/types.ts
+++ b/ui/frontend/types.ts
@@ -66,3 +66,8 @@ export enum Edition {
   Rust2015 = '2015',
   Rust2018 = '2018',
 }
+
+export enum Backtrace {
+  Disabled = 'disabled',
+  Enabled = 'enabled',
+}

--- a/ui/src/main.rs
+++ b/ui/src/main.rs
@@ -525,6 +525,7 @@ struct CompileRequest {
     #[serde(rename = "crateType")]
     crate_type: String,
     tests: bool,
+    backtrace: bool,
     code: String,
 }
 
@@ -545,6 +546,7 @@ struct ExecuteRequest {
     #[serde(rename = "crateType")]
     crate_type: String,
     tests: bool,
+    backtrace: bool,
     code: String,
 }
 
@@ -657,6 +659,7 @@ impl TryFrom<CompileRequest> for sandbox::CompileRequest {
             edition: parse_edition(&me.edition)?,
             crate_type: parse_crate_type(&me.crate_type)?,
             tests: me.tests,
+            backtrace: me.backtrace,
             code: me.code,
         })
     }
@@ -683,6 +686,7 @@ impl TryFrom<ExecuteRequest> for sandbox::ExecuteRequest {
             edition: parse_edition(&me.edition)?,
             crate_type: try!(parse_crate_type(&me.crate_type)),
             tests: me.tests,
+            backtrace: me.backtrace,
             code: me.code,
         })
     }
@@ -779,6 +783,7 @@ impl TryFrom<EvaluateRequest> for sandbox::ExecuteRequest {
             edition: None, // FIXME: What should this be?
             crate_type: sandbox::CrateType::Binary,
             tests: false,
+            backtrace: false,
             code: me.code,
         })
     }

--- a/ui/src/sandbox.rs
+++ b/ui/src/sandbox.rs
@@ -113,7 +113,7 @@ impl Sandbox {
     pub fn compile(&self, req: &CompileRequest) -> Result<CompileResponse> {
         try!(self.write_source_code(&req.code));
 
-        let mut command = self.compile_command(req.target, req.channel, req.mode, req.crate_type, req.tests, req.edition);
+        let mut command = self.compile_command(req.target, req.channel, req.mode, req.crate_type, req.tests, req.backtrace, req.edition);
 
         let output = try!(command.output().map_err(Error::UnableToExecuteCompiler));
 
@@ -167,7 +167,7 @@ impl Sandbox {
 
     pub fn execute(&self, req: &ExecuteRequest) -> Result<ExecuteResponse> {
         try!(self.write_source_code(&req.code));
-        let mut command = self.execute_command(req.channel, req.mode, req.crate_type, req.tests, req.edition);
+        let mut command = self.execute_command(req.channel, req.mode, req.crate_type, req.tests, req.backtrace, req.edition);
 
         let output = try!(command.output().map_err(Error::UnableToExecuteCompiler));
 
@@ -257,9 +257,9 @@ impl Sandbox {
         Ok(())
     }
 
-    fn compile_command(&self, target: CompileTarget, channel: Channel, mode: Mode, crate_type: CrateType, tests: bool, edition: Option<Edition>) -> Command {
+    fn compile_command(&self, target: CompileTarget, channel: Channel, mode: Mode, crate_type: CrateType, tests: bool, backtrace: bool, edition: Option<Edition>) -> Command {
         let mut cmd = self.docker_command(Some(crate_type));
-        set_execution_environment(&mut cmd, Some(target), crate_type, edition);
+        set_execution_environment(&mut cmd, Some(target), crate_type, edition, backtrace);
 
         let execution_cmd = build_execution_command(Some(target), mode, crate_type, tests);
 
@@ -270,9 +270,9 @@ impl Sandbox {
         cmd
     }
 
-    fn execute_command(&self, channel: Channel, mode: Mode, crate_type: CrateType, tests: bool, edition: Option<Edition>) -> Command {
+    fn execute_command(&self, channel: Channel, mode: Mode, crate_type: CrateType, tests: bool, backtrace: bool, edition: Option<Edition>) -> Command {
         let mut cmd = self.docker_command(Some(crate_type));
-        set_execution_environment(&mut cmd, None, crate_type, edition);
+        set_execution_environment(&mut cmd, None, crate_type, edition, backtrace);
 
         let execution_cmd = build_execution_command(None, mode, crate_type, tests);
 
@@ -392,7 +392,7 @@ fn build_execution_command(target: Option<CompileTarget>, mode: Mode, crate_type
     cmd
 }
 
-fn set_execution_environment(cmd: &mut Command, target: Option<CompileTarget>, crate_type: CrateType, edition: Option<Edition>) {
+fn set_execution_environment(cmd: &mut Command, target: Option<CompileTarget>, crate_type: CrateType, edition: Option<Edition>, backtrace: bool) {
     use self::CompileTarget::*;
     use self::CrateType::*;
 
@@ -407,6 +407,10 @@ fn set_execution_environment(cmd: &mut Command, target: Option<CompileTarget>, c
 
     if let Some(edition) = edition {
         cmd.args(&["--env", &format!("PLAYGROUND_EDITION={}", edition.cargo_ident())]);
+    }
+
+    if backtrace {
+        cmd.args(&["--env", "RUST_BACKTRACE=1"]);
     }
 }
 
@@ -565,6 +569,7 @@ pub struct CompileRequest {
     pub mode: Mode,
     pub edition: Option<Edition>,
     pub tests: bool,
+    pub backtrace: bool,
     pub code: String,
 }
 
@@ -583,6 +588,7 @@ pub struct ExecuteRequest {
     pub edition: Option<Edition>,
     pub crate_type: CrateType,
     pub tests: bool,
+    pub backtrace: bool,
     pub code: String,
 }
 
@@ -637,6 +643,7 @@ mod test {
                 tests: false,
                 code: HELLO_WORLD_CODE.to_string(),
                 edition: None,
+                backtrace: false,
             }
         }
     }
@@ -651,6 +658,7 @@ mod test {
                 tests: false,
                 code: HELLO_WORLD_CODE.to_string(),
                 edition: None,
+                backtrace: false,
             }
         }
     }
@@ -816,6 +824,50 @@ mod test {
         let resp = sb.execute(&req)?;
 
         assert!(!resp.stderr.contains("`crate` in paths is experimental"));
+        Ok(())
+    }
+
+    const BACKTRACE_CODE: &str = r#"
+    fn trigger_the_problem() {
+        None::<u8>.unwrap();
+    }
+
+    fn main() {
+        trigger_the_problem()
+    }
+    "#;
+
+    #[test]
+    fn backtrace_disabled() -> Result<()> {
+        let req = ExecuteRequest {
+            code: BACKTRACE_CODE.to_string(),
+            backtrace: false,
+            ..ExecuteRequest::default()
+        };
+
+        let sb = Sandbox::new()?;
+        let resp = sb.execute(&req)?;
+
+        assert!(resp.stderr.contains("Run with `RUST_BACKTRACE=1` for a backtrace"));
+        assert!(!resp.stderr.contains("stack backtrace:"));
+
+        Ok(())
+    }
+
+    #[test]
+    fn backtrace_enabled() -> Result<()> {
+        let req = ExecuteRequest {
+            code: BACKTRACE_CODE.to_string(),
+            backtrace: true,
+            ..ExecuteRequest::default()
+        };
+
+        let sb = Sandbox::new()?;
+        let resp = sb.execute(&req)?;
+
+        assert!(!resp.stderr.contains("Run with `RUST_BACKTRACE=1` for a backtrace"));
+        assert!(resp.stderr.contains("stack backtrace:"));
+
         Ok(())
     }
 

--- a/ui/src/sandbox.rs
+++ b/ui/src/sandbox.rs
@@ -762,6 +762,63 @@ mod test {
         assert!(resp.stdout.contains("nightly"));
     }
 
+    const EDITION_CODE: &str = r#"
+    mod foo {
+        pub fn bar() {}
+    }
+
+    fn main() {
+        crate::foo::bar();
+    }
+    "#;
+
+    #[test]
+    fn rust_edition_default() -> Result<()> {
+        let req = ExecuteRequest {
+            channel: Channel::Nightly,
+            code: EDITION_CODE.to_string(),
+            ..ExecuteRequest::default()
+        };
+
+        let sb = Sandbox::new()?;
+        let resp = sb.execute(&req)?;
+
+        assert!(resp.stderr.contains("`crate` in paths is experimental"));
+        Ok(())
+    }
+
+    #[test]
+    fn rust_edition_2015() -> Result<()> {
+        let req = ExecuteRequest {
+            channel: Channel::Nightly,
+            code: EDITION_CODE.to_string(),
+            edition: Some(Edition::Rust2015),
+            ..ExecuteRequest::default()
+        };
+
+        let sb = Sandbox::new()?;
+        let resp = sb.execute(&req)?;
+
+        assert!(resp.stderr.contains("`crate` in paths is experimental"));
+        Ok(())
+    }
+
+    #[test]
+    fn rust_edition_2018() -> Result<()> {
+        let req = ExecuteRequest {
+            channel: Channel::Nightly,
+            code: EDITION_CODE.to_string(),
+            edition: Some(Edition::Rust2018),
+            ..ExecuteRequest::default()
+        };
+
+        let sb = Sandbox::new()?;
+        let resp = sb.execute(&req)?;
+
+        assert!(!resp.stderr.contains("`crate` in paths is experimental"));
+        Ok(())
+    }
+
     #[test]
     fn output_llvm_ir() {
         let req = CompileRequest {

--- a/ui/src/sandbox.rs
+++ b/ui/src/sandbox.rs
@@ -628,16 +628,36 @@ mod test {
     }
     "#;
 
+    impl Default for ExecuteRequest {
+        fn default() -> Self {
+            ExecuteRequest {
+                channel: Channel::Stable,
+                crate_type: CrateType::Binary,
+                mode: Mode::Debug,
+                tests: false,
+                code: HELLO_WORLD_CODE.to_string(),
+                edition: None,
+            }
+        }
+    }
+
+    impl Default for CompileRequest {
+        fn default() -> Self {
+            CompileRequest {
+                target: CompileTarget::LlvmIr,
+                channel: Channel::Stable,
+                crate_type: CrateType::Binary,
+                mode: Mode::Debug,
+                tests: false,
+                code: HELLO_WORLD_CODE.to_string(),
+                edition: None,
+            }
+        }
+    }
+
     #[test]
     fn basic_functionality() {
-        let req = ExecuteRequest {
-            channel: Channel::Stable,
-            crate_type: CrateType::Binary,
-            mode: Mode::Debug,
-            tests: false,
-            code: HELLO_WORLD_CODE.to_string(),
-            edition: None,
-        };
+        let req = ExecuteRequest::default();
 
         let sb = Sandbox::new().expect("Unable to create sandbox");
         let resp = sb.execute(&req).expect("Unable to execute code");
@@ -660,12 +680,8 @@ mod test {
     #[test]
     fn debug_mode() {
         let req = ExecuteRequest {
-            channel: Channel::Stable,
-            crate_type: CrateType::Binary,
-            mode: Mode::Debug,
-            tests: false,
             code: COMPILATION_MODE_CODE.to_string(),
-            edition: None,
+            ..ExecuteRequest::default()
         };
 
         let sb = Sandbox::new().expect("Unable to create sandbox");
@@ -677,12 +693,9 @@ mod test {
     #[test]
     fn release_mode() {
         let req = ExecuteRequest {
-            channel: Channel::Stable,
-            crate_type: CrateType::Binary,
             mode: Mode::Release,
-            tests: false,
             code: COMPILATION_MODE_CODE.to_string(),
-            edition: None,
+            ..ExecuteRequest::default()
         };
 
         let sb = Sandbox::new().expect("Unable to create sandbox");
@@ -705,11 +718,8 @@ mod test {
     fn stable_channel() {
         let req = ExecuteRequest {
             channel: Channel::Stable,
-            crate_type: CrateType::Binary,
-            mode: Mode::Debug,
-            tests: false,
             code: VERSION_CODE.to_string(),
-            edition: None,
+            ..ExecuteRequest::default()
         };
 
         let sb = Sandbox::new().expect("Unable to create sandbox");
@@ -724,11 +734,8 @@ mod test {
     fn beta_channel() {
         let req = ExecuteRequest {
             channel: Channel::Beta,
-            crate_type: CrateType::Binary,
-            mode: Mode::Debug,
-            tests: false,
             code: VERSION_CODE.to_string(),
-            edition: None,
+            ..ExecuteRequest::default()
         };
 
         let sb = Sandbox::new().expect("Unable to create sandbox");
@@ -743,11 +750,8 @@ mod test {
     fn nightly_channel() {
         let req = ExecuteRequest {
             channel: Channel::Nightly,
-            crate_type: CrateType::Binary,
-            mode: Mode::Debug,
-            tests: false,
             code: VERSION_CODE.to_string(),
-            edition: Some(Edition::Rust2018),
+            ..ExecuteRequest::default()
         };
 
         let sb = Sandbox::new().expect("Unable to create sandbox");
@@ -762,12 +766,7 @@ mod test {
     fn output_llvm_ir() {
         let req = CompileRequest {
             target: CompileTarget::LlvmIr,
-            channel: Channel::Stable,
-            crate_type: CrateType::Binary,
-            mode: Mode::Debug,
-            tests: false,
-            code: HELLO_WORLD_CODE.to_string(),
-            edition: None,
+            ..CompileRequest::default()
         };
 
         let sb = Sandbox::new().expect("Unable to create sandbox");
@@ -782,12 +781,7 @@ mod test {
     fn output_assembly() {
         let req = CompileRequest {
             target: CompileTarget::Assembly(AssemblyFlavor::Att, DemangleAssembly::Mangle, ProcessAssembly::Raw),
-            channel: Channel::Stable,
-            crate_type: CrateType::Binary,
-            mode: Mode::Debug,
-            tests: false,
-            code: HELLO_WORLD_CODE.to_string(),
-            edition: None,
+            ..CompileRequest::default()
         };
 
         let sb = Sandbox::new().expect("Unable to create sandbox");
@@ -803,12 +797,7 @@ mod test {
     fn output_demangled_assembly() {
         let req = CompileRequest {
             target: CompileTarget::Assembly(AssemblyFlavor::Att, DemangleAssembly::Demangle, ProcessAssembly::Raw),
-            channel: Channel::Stable,
-            crate_type: CrateType::Binary,
-            mode: Mode::Debug,
-            tests: false,
-            code: HELLO_WORLD_CODE.to_string(),
-            edition: None,
+            ..CompileRequest::default()
         };
 
         let sb = Sandbox::new().expect("Unable to create sandbox");
@@ -823,12 +812,7 @@ mod test {
     fn output_filtered_assembly() {
         let req = CompileRequest {
             target: CompileTarget::Assembly(AssemblyFlavor::Att, DemangleAssembly::Mangle, ProcessAssembly::Filter),
-            channel: Channel::Stable,
-            crate_type: CrateType::Binary,
-            mode: Mode::Debug,
-            tests: false,
-            code: HELLO_WORLD_CODE.to_string(),
-            edition: None,
+            ..CompileRequest::default()
         };
 
         let sb = Sandbox::new().expect("Unable to create sandbox");
@@ -886,12 +870,8 @@ mod test {
         "#;
 
         let req = ExecuteRequest {
-            channel: Channel::Stable,
-            mode: Mode::Debug,
-            crate_type: CrateType::Binary,
-            tests: false,
             code: code.to_string(),
-            edition: None
+            ..ExecuteRequest::default()
         };
 
         let sb = Sandbox::new().expect("Unable to create sandbox");
@@ -911,12 +891,8 @@ mod test {
         "#;
 
         let req = ExecuteRequest {
-            channel: Channel::Stable,
-            mode: Mode::Debug,
-            crate_type: CrateType::Binary,
-            tests: false,
             code: code.to_string(),
-            edition: None,
+            ..ExecuteRequest::default()
         };
 
         let sb = Sandbox::new().expect("Unable to create sandbox");
@@ -935,12 +911,8 @@ mod test {
         "#;
 
         let req = ExecuteRequest {
-            channel: Channel::Stable,
-            mode: Mode::Debug,
-            crate_type: CrateType::Binary,
-            tests: false,
             code: code.to_string(),
-            edition: None,
+            ..ExecuteRequest::default()
         };
 
         let sb = Sandbox::new().expect("Unable to create sandbox");
@@ -964,12 +936,8 @@ mod test {
         "##;
 
         let req = ExecuteRequest {
-            channel: Channel::Stable,
-            mode: Mode::Debug,
-            crate_type: CrateType::Binary,
-            tests: false,
             code: forkbomb.to_string(),
-            edition: None,
+            ..ExecuteRequest::default()
         };
 
         let sb = Sandbox::new().expect("Unable to create sandbox");

--- a/ui/src/sandbox.rs
+++ b/ui/src/sandbox.rs
@@ -886,7 +886,7 @@ mod test {
             fn main() {
                 let megabyte = 1024 * 1024;
                 let mut big = vec![0u8; 384 * megabyte];
-                *big.last_mut().unwrap() += 1;
+                for i in &mut big { *i += 1; }
             }
         "#;
 

